File: blink-windows/src/Blink.Core/Timer/TimerStateMachine.cs
using Blink.Core.FlowDetection;

namespace Blink.Core.Timer;

/// <summary>
/// 20-minute countdown timer. Pauses during idle/meeting.
/// </summary>
public sealed class TimerStateMachine
{
    public const double DefaultDuration = 1200; // 20 minutes

    public double RemainingSeconds { get; private set; } = DefaultDuration;
    public bool IsPaused { get; private set; }
    public event Action? OnBreakDue;

    // Tracks the duration of the *current* cycle. Set by Reset(duration) so
    // that the UI's progress bar starts at 0% when an extension begins.
    public double TimerDuration { get; private set; } = DefaultDuration;

    public double Progress
    {
        get
        {
            return TimerDuration > 0
                ? Math.Max(0, Math.Min(1.0, 1.0 - RemainingSeconds / TimerDuration))
                : 1.0;
        }
    }

    public void Tick(FlowState flowState, double deltaSeconds)
    {
        if (flowState is FlowState.Idle or FlowState.Meeting)
        {
            IsPaused = true;
            return;
        }

        IsPaused = false;

        if (flowState == FlowState.BreakPrompted) return;
        if (RemainingSeconds <= 0) return;

        RemainingSeconds -= deltaSeconds;

        if (RemainingSeconds <= 0)
        {
            RemainingSeconds = 0;
            OnBreakDue?.Invoke();
        }
    }

    public void ResetAfterBreak()
    {
        RemainingSeconds = DefaultDuration;
        TimerDuration = DefaultDuration;
        IsPaused = false;
    }

    public void Reset(double duration)
    {
        RemainingSeconds = duration;
        TimerDuration = duration;
        IsPaused = false;
    }
}

File: blink-macos/BlinkCore/Sources/BlinkCore/Timer/TimerStateMachine.swift
import Foundation

/// Manages the break countdown timer.
/// Counts down every second. Pauses when idle or in a meeting.
public final class TimerStateMachine {
    public static let defaultDuration: TimeInterval = 1200  // 20 minutes

    public private(set) var remainingSeconds: TimeInterval
    public private(set) var isPaused: Bool = false

    public var onBreakDue: (() -> Void)?

    public init() {
        self.remainingSeconds = Self.defaultDuration
    }

    /// Total duration for progress bar calculation.
    public var timerDuration: TimeInterval { Self.defaultDuration }

    /// Progress from 0.0 (just started) to 1.0 (break due).
    public var progress: Double {
        guard Self.defaultDuration > 0 else { return 1.0 }
        return max(0, min(1.0, 1.0 - remainingSeconds / Self.defaultDuration))
    }

    /// Called every second to advance the timer.
    public func tick(flowState: FlowState, deltaSeconds: TimeInterval) {
        if flowState == .idle || flowState == .meeting {
            isPaused = true
            return
        }

        isPaused = false

        guard flowState != .breakPrompted else { return }
        guard remainingSeconds > 0 else { return }

        remainingSeconds -= deltaSeconds

        if remainingSeconds <= 0 {
            remainingSeconds = 0
            onBreakDue?()
        }
    }

    /// Reset timer after a break is taken.
    public func resetAfterBreak() {
        remainingSeconds = Self.defaultDuration
        isPaused = false
    }

    /// Reset with a specific duration (for extensions).
    public func reset(duration: TimeInterval) {
        remainingSeconds = duration
        isPaused = false
    }
}
